[NAS] 미션 2: 체스판 생성

Malloc72P opened the pull request:
안녕하세요? 백엔드 클래스의 NAS라고 해요!

미션 2를 완료해서 pull request를 남기게 되었어요!

이번에  변경한 내용을 요약해보았습니다!
* assertEquals 대신에 AssertJ를 사용하도록 수정했어요. 확실히 assertEquals보다 assertThat을 쓰는게 가독성이 더 좋은것 같네요!
* 커밋 메세지를 예전보다 더 잘쓰려고 노력했어요! 근데 역시나 영어로 쓰는건 아직은 무리인 것 같네요. 😢 대신 요약메세지는 명령조로 작성했어요! 그리고 본문과 요약메세지 사이에 공백라인을 추가했어요! 😄 
* 예전에 체스말을 구현하려고 추상클래스를 썼는데, 이번에는 지웠어요. 나중에 코드가 중복되는일이 생기면, 그때가서 추상클래스로 뽑아도 늦지 않을 것 같네요! 🥰 
* 코드 포매팅을 일관성있게 해보려고 노력했어요! Ctrl + Alt + L을 누르는 습관을 들이고, Jetbrains에서도 커밋하기 전에 리포매팅 하는 설정을 켰어요! 좀 더 보기 예쁜 코드가 되었으면 좋겠어요! 👍 
* 그 외에는 최대한 요구사항에 맞게 작성하려고 노력했어요! 혹시나 요구사항을 지키지 못한 부분이 있다면, 피드백 주시면 정말 감사하겠습니다! 

이상 글을 마치겠습니다! 제 글을 읽어주셔서 정말 감사합니다! 👍

Malloc72P committed:
Use AssertJ instead of assertEquals

Malloc72P committed:
Reformat your code

Malloc72P committed:
Correct the access modifiers

Malloc72P committed:
PawnTest의 테스트 메서드 이름을 고쳐라

좀 더 알아보기 쉬운 이름으로 수정함

Malloc72P committed:
색깔을 상수로 쓸 수 있도록 바꿔라

- enum으로 쓸 수 있는 색깔을 선언함
- 색깔을 생성자에 넣을때 enum을 써서 넣도록 수정함

Malloc72P committed:
Pawn클래스에 매개변수 없는 생성자를 추가해라

Malloc72P committed:
ChessPiece 추상클래스를 제거하라

- 추상클래스에 있던 내용은 Pawn클래스에 옮김.

Malloc72P committed:
체스판 클래스를 구현하라

- 체스판에 폰을 추가하고 찾는 메서드 구현함
- 체스판을 테스트하는 코드 추가함

Malloc72P committed:
보드에 추가된 폰의 개수를 리턴하는 메서드를 구현하라

- Board클래스에 size라는 메서드를 추가해서 위의 요구사항을 만족시켰음.
- size메서드를 테스트하는 코드를 추가하였음

Malloc72P committed:
테스트 코드의 패키지를 분리하라

- 폰과 보드 테스트 클래스를 chess라는 패키지로 묶고, 전부 실행해서 테스트를 통과하는걸 확인했다.

Malloc72P committed:
체스판에 폰 이외의 객체가 추가되는지 테스트하라

- 폰 이외의 타입이 추가되면 예외가 발생하는지 테스트하는 코드 추가함
- null이 추가될 수 있는지 검사하는 테스트 코드 추가함.
- 그 결과 null이 추가될 수 있는 문제를 식별함. add메서드에서 null이 추가되면 예외를 발생시키도록 수정함

Malloc72P committed:
패키지를 요구사항에 맞게 다시 분리하라

- pieces랑 chess 패키지로 분리했어야 하는데 그렇게 않했음. 그래서 다시 분리함.

Malloc72P committed:
BoardTest에서 중복코드를 제거하라

- BeforeEach 어노테이션을 써서 테스트용 보드를 생성하도록 수정함
- 추가 및 찾기를 검사하는 testAdditionAndFind메서드에서 중복코드를 메서드로 추출하였음.

SANGCO commented on src/test/java/pieces/PawnTest.java:
```diff
@@ -0,0 +1,30 @@
+    @Test
+    @DisplayName("생성자에 넣은 색깔대로 폰이 생성되어야 합니다.")
+    void testColorOfPawn() {
+        for (ColorOfChessPiece tc : ColorOfChessPiece.values()) {
```
enum values() 사용하는거 좋네요~

Malloc72P replied:
감사합니다! enum 진짜 좋은 것 같아요!

SANGCO commented on src/test/java/chess/BoardTest.java:
```diff
@@ -0,0 +1,57 @@
+
+    void verifyAdditionAndFind(Pawn pawn, int idx) {
+        board.add(pawn);
+        assertThat(board.findPawn(idx)).isEqualTo(pawn);
```
assertAll()에 대해서 학습하고 적용해 보는건 어떨까요~
[JUnit 5 소개](https://javacan.tistory.com/entry/JUnit-5-Intro) (링크에 assertAll() 부분 참고해 주세요.)

Malloc72P replied:
```java
  for (int i = 0; i < testcases.length; i++) {
      final int idx = i;
      assertAll(
              () -> verifyAddition(testcases[idx], idx),
              () -> verifyFind(testcases[idx], idx)
      );
  }
```
assertAll을 적용했어요! 😄 
예전에는 verifyAdditionAndFind메서드에서 보드에 폰을 추가하고, 찾는 기능을 테스트하다보니까, 추가하는거에서 실패하면 검색하는 기능은 테스트하지 않았는데, assertAll을 쓰니까 폰 추가하기 기능에서 실패해도 검색기능을 테스트할 수 있어서 좋은 것 같아요! 😮 
좋은 기능을 알려주셔서 감사합니다!

SANGCO commented on src/test/java/chess/BoardTest.java:
```diff
@@ -0,0 +1,57 @@
+    @Test
+    @DisplayName("체스 보드에 폰을 추가하고, 찾을 수 있어야 합니다.")
+    void testAdditionAndFind() {
+        Pawn[] testcases = {
+                new Pawn(),
+                new Pawn(ColorOfChessPiece.BLACK)
+        };
+        for (int i = 0; i < testcases.length; i++)
+            verifyAdditionAndFind(testcases[i], i);
```
깔끔하고 좋네요~

Malloc72P replied:
감사합니다~ 다른 코드도 깔끔하게 작성할 수 있도록 노력하겠습니다~ 😄

MJbae replied:
이렇게도 테스트를 할 수 있군요! 나스에게 많이 배워갑니다~

SANGCO commented on src/test/java/chess/BoardTest.java:
```diff
@@ -0,0 +1,57 @@
+                7,
+                "Error"
+        };
+        for (Object tc : testcases)
+            assertThatThrownBy(() -> {
+                board.add((Pawn) tc);
+            }).isInstanceOf(ClassCastException.class);
+        assertThatThrownBy(() -> {
+            board.add(null);
+        }).isInstanceOf(InvalidParameterException.class);
```
이 부분은 뭔가 코드가 지저분해 보이네요.
위에 verifyAdditionAndFind() 메서드 처럼 메서드 분리를 하면 어떨까요?

```java
private void test(Object o, Class<?> c) {
    assertThatThrownBy(() -> {}).isInstanceOf(c);
}
```
뭐 대충 이런식으로 위에서 하셨던거 처럼 메서드 분리를 시도해 보시면 좋을거 같습니다.

Malloc72P replied:
```java
for (Object tc : testcases)
    testAdditionThrowCastException(tc);
testAdditionThrowInvalidParameterException();
```
메서드를 분리했어요! 😄 
캐스팅 익셉션이 발생하는 테스트케이스는 testAdditionThrowCastException메서드로 하고, null같은걸 아규먼트로 넘기는 경우엔 testAdditionThrowInvalidParameterException메서드로 테스트하도록 수정했어요!
각각의 테스트메서드는 아래와 같습니다!
```java
void testAdditionThrowCastException(Object tc) {
    assertThatThrownBy(() -> {
        board.add((Pawn) tc);
    }).isInstanceOf(ClassCastException.class);
}

void testAdditionThrowInvalidParameterException() {
    assertThatThrownBy(() -> {
        board.add(null);
    }).isInstanceOf(InvalidParameterException.class);
}
```

SANGCO replied:
Nas님 메소드 분리 잘하셨네요~
근데 위에서 피드백 드렸던거 처럼 특정 익셉션도 파라미터로 넘기는 방식으로
testAdditionThrowCastException(), testAdditionThrowInvalidParameterException() 메서드를
합치는것도 한번 생각해 보아요~

Malloc72P replied:
앗 코멘트 달아주셨는데 이 부분을 놓쳤네요! 😮 

말씀해주신대로 적용해봤어요!

```java
for (Object tc : testcases)
    testAdditionThrowException(tc, ClassCastException.class);
testAdditionThrowException(null, InvalidParameterException.class);

void testAdditionThrowException(Object tc, Class<?> exceptionClass) {
    assertThatThrownBy(() -> {
        board.add((Pawn) tc);
    }).isInstanceOf(exceptionClass);
}
```
testAdditionThrowException라는 메서드로 합쳤어요! 그리고 두번째 패러미터를 코멘트로 달아주신 것처럼 Class<?>를 써서, 원하는 예외가 발생했는지 검사하도록 수정했어요! 정말 많이 배워갑니다~! 😄

SANGCO commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    public void add(Pawn piece) {
+        if (piece == null)
```
에외 처리 좋네요~

Malloc72P replied:
감사합니다! 다른 메서드도 이렇게 예외처리하는 코드를 작성하면 좋을 것 같아요!

SANGCO commented on src/main/java/net/nas/chess/Board.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    public Pawn findPawn(int idx) {
+        return pieceList.stream()
+                .skip(idx)
+                .findFirst()
+                .orElse(null);
```
인덱스로 pawn을 찾는데 stream을 쓸 필요가 있었을까요?

우선 이 코드에 대한 테스트 코드를 작성해 봅시다.

어떻게 동작 하기를 기대하고 코딩을 했고
테스트 결과 코드가 의도한데로 잘 동작 했는지를 댓글로 알려주세요~

Malloc72P replied:
코멘트를 받고, 다시 제 코드를 보니까 실수했다는걸 알게 되었어요 😆 
pieceList가 ArrayList니까 그냥 pieceList.get(idx)를 하면 되는거였는데 불필요하게 스트림을 쓴 것 같아요. 🤣 

그리고 findPawn메서드에서도 검사를 해주면 좋겠더라구요.
```java
public Pawn findPawn(int idx) {
    if (idx < 0 || pieceList.size() <= idx)
        throw new InvalidParameterException("index exceeded the bounds of the pieceList");
    return pieceList.get(idx);
}
```
그래서 idx가 0보다 작거나,  size랑 같거나 크면 인덱스가 배열의 범위를 벗어난거라서 예외를 발생시키도록 수정했어요!
이 예외가 실제로 발생하는지를 테스트하는 메서드도 추가로 작성했어요! 😄 
```java
@Test
@DisplayName("폰을 찾을때 넣는 인덱스가 배열의 범위를 벗어나면 예외가 발생해야 합니다")
void testErrorFind() {
    testFindThrowInvalidParameterException(-1);
    testFindThrowInvalidParameterException(0);
    for (int i = 0; i < 2; i++) {
        board.add(new Pawn());
        testFindThrowInvalidParameterException(i + 1);
    }
}
void testFindThrowInvalidParameterException(int idx) {
    assertThatThrownBy(() -> {
        board.findPawn(idx);
    }).isInstanceOf(InvalidParameterException.class);
}
```

Malloc72P replied:
testFindThrowInvalidParameterException메서드도 수정했어요!
이 메서드도 패러미터로 검사하고자 하는 예외의 클래스를 받도록 수정해서, 나중에 다른 예외상황을 테스트할때 재사용할 수 있도록 했어요! 😄 
```java
void testFindThrowException(int idx, Class<?> exceptionClass) {
    assertThatThrownBy(() -> {
        board.findPawn(idx);
    }).isInstanceOf(exceptionClass);
}
```

SANGCO reviewed (commented):
NAS님 안녕하세요, 리뷰어 상코입니다.
작성하신 코드에 대해서 코멘트를 남겼습니다. 
확인해 보시고 질문 있으시면 댓글 부탁드립니다~

SANGCO commented on src/test/java/chess/BoardTest.java:
```diff
@@ -0,0 +1,57 @@
+package chess;
+
+import net.nas.chess.Board;
```
소스코드와 테스트코드 패키지가 다르네요.
그래서 Board를 테스트 하는데 Board를 import 하고 있군요.
소스코드와 테스트코드 패키지를 동일한 구조로 만들어 주세요~

그리고 추가로 java.net.honux 패키지는 지우는게 좋을거 같네요~

Malloc72P replied:
와 이거 이렇게 하는 거였군요? 😮 
코멘트 달아주신대로 소스코드랑 테스트코드의 패키지구조를 똑같게 만들어주니까, BoardTest는 Board를 import하지 않아도 되네요! 😄 
알려주셔서 정말 감사드려요! 👍 

아 그리고 honux패키지를 제가 지워도 괜찮을까요? 비록 안쓰이는 소스코드이긴 한데, 다른 사람이 작성한거라서 지워도 되는지 조금 조심스럽네요...

SANGCO replied:
이게 협업을 하는 프로젝트가 아니라서 쓸데없이 안쓰는 패키지는 깔끔하게 지우는게 좋을거 같아요~

SANGCO reviewed (commented):
리뷰 추가합니다~

Malloc72P committed:
assertAll을 적용해라

testAdditionAndFind메서드에서 assertAll을 쓰도록 수정함

Malloc72P committed:
testErrorAddition메서드를 리팩토링하라

- 메서드를 분리해서 가독성을 향상시키기 위한 노력을 했음

Malloc72P committed:
Pawn클래스의 findPawn메서드를 개선하라

- ArrayList라서 그냥 get()메서드를 쓰면 원하는 인덱스의 폰을 찾을 수 있는데, 불필요하게 스트림을 쓴 부분을 수정함
- 추가로 인덱스가 배열의 범위를 벗어나면 예외를 발생시키도록 수정함

Malloc72P committed:
Board에서 폰을 찾는 기능에서 예외처리할 수 있도록 수정해라

- pieceList라는 ArrayList의 범위를 벗어나는 인덱스가 패러미터로 넘어오면 예외를 발생시키도록 수정함

Malloc72P committed:
Board클래스에서 find를 잘못해서 예외를 발생시키는 상황을 테스트하라

Malloc72P committed:
소스코드랑 테스트코드의 패키지구조를 똑같게 만들어라

- 패키지 구조를 똑같게 만드니까, BoardTest에서는 Board를 import하지 않아도 쓸 수 있게 되었다.
- PawnTest도 마찬가지.

Malloc72P committed:
testAdditionThrowCastException메서드와 invalidParameterException메서드를 합쳐라

- testAdditionThrowException메서드로 합쳤음. 두번째 패러미터로 검사하고자 하는 예외의 클래스를 받도록 했음.

Malloc72P committed:
testFindThrowInvalidParameterException메서드의 재사용성을 높여라

- testFindThrowException메서드로 이름을 바꾸고, 패러미터로 검사하고자 하는 예외의 클래스를 받아서 검사하도록 해서 재사용성을 높였음

Malloc72P commented:
코멘트 달아주신걸 보고, 수정해서 커밋했어요!

그런데 아직 honux 패키지는 지우지 못했어요. 😢 
다른 사람이 작성했던거라 쉽게 지워도 되는건지 조금 조심스럽네요...
이럴땐 어떻게 하는게 좋을까요? 작성자한테 말하고 지우는게 좋을까요?
아니면 어차피 제 브랜치니까 그냥 지워도 괜찮을까요? 🍉

Malloc72P committed:
안쓰는 패키지를 제거하라

- net.honux 패키지는 사용하지 않기 때문에 삭제함

Malloc72P commented:
안쓰는 패키지까지 제거완료했습니다! 😃

SANGCO commented:
NAS님 수정 너무 잘하셨네요~
간단한거 몇가지만 더 수정하고 넘어갑시다.

테스트 클래스에서 클래스와 메서드에 public을 생략하는건 좋은데
테스트 클래스 내부에서만 쓰는 메서드는 앞에 private을 붙여주는게 좋을거 같아요.

그리고 BoardTest 클래스에 보면 verifyAddition(), verifyFind() 메서드가
testAdditionAndFind() 메서드에서 쓰고 있는데 testErrorFind() 메서드 밑에 위치하고 있는데요.
저 두 개의 메서드를 직접 사용하고 있는 testAdditionAndFind() 메서드 밑으로 옮기면 코드를 읽을 때 더 좋지 않을까요~

Malloc72P committed:
테스트 클래스의 접근제어자를 고치고, 메서드 위치를 조정하라

- 테스트 클래스 내부에서만 쓰이는 메서드는 private으로 선언하도록 수정함
- 가독성을 향상시키기 위해, 메서드가 작성된 위치를 조정함

Malloc72P commented:
테스트 클래스에 있는 메서드의 접근제어자를 고쳤어요! 확실히 테스트 클래스 내부에서만 쓰이는 메서드는 private으로 써주는게 더 좋은 것 같아요! 😉 

그리고 verifyAddition이랑 verifyFind메서드의 위치를 testAdditionAndFind메서드의 밑으로 옮기니까 좀 더 보기 좋은 것 같아요! 뭔가 제목을 읽은다음 본문을 읽는 느낌이 들어서 좋은 것 같네요! 😃

SANGCO commented on src/test/java/net/nas/chess/BoardTest.java:
```diff
@@ -0,0 +1,85 @@
+
+
+public class BoardTest {
+    Board board;
```
BoardTest 클래스에서만 사용하는 Board 클래스 접근 제한자는 private이 좋겠죠?

Malloc72P replied:
앗 자꾸 접근제한자를 빠뜨리네요! 😮 주의하겠습니다!
다음 미션 시작할때 이 부분부터 수정해야겠네요!
알려주셔서 감사합니다!

SANGCO reviewed (approved):
수정 잘하셨네요~
다음 미션도 화이팅 화이팅

honux77 commented:
상코 NAS 다 수고 많으셨어요!
